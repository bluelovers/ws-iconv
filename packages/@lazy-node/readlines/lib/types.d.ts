/// <reference types="node" />
/// <reference types="node" />
/**
 * Created by user on 2020/5/29.
 */
import { PathLike } from 'fs';
export interface IOptions {
    readChunk?: number;
    newLineCharacter?: INewLineCharacter;
}
export declare const enum EnumNewLineCharacter {
    LF = 10,
    CR = 13
}
export declare type IPathLike = PathLike | number;
export declare type INewLineCharacter = string | number | EnumNewLineCharacter | (string | number | EnumNewLineCharacter)[] | Buffer;
