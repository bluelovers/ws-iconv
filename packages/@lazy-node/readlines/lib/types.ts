/**
 * Created by user on 2020/5/29.
 */
import { PathLike } from 'fs';

export interface IOptions
{
	readChunk?: number;
	newLineCharacter?: INewLineCharacter;
}

export const enum EnumNewLineCharacter
{
	LF = 0x0a,
	CR = 0x0d,
}

export type IPathLike = PathLike | number
export type INewLineCharacter = string | number | EnumNewLineCharacter | (string | number | EnumNewLineCharacter)[] | Buffer

