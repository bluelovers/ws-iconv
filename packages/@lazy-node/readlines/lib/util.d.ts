/// <reference types="node" />
import { INewLineCharacter } from './types';
export declare function handleOptionNewLineCharacter(newLineCharacter?: INewLineCharacter): number[];
export declare function bufferEndWithByBuffer(buf: Buffer, value: Uint8Array, byteOffset?: number, encoding?: BufferEncoding): boolean;
export declare function bufferEndWith(buf: Buffer, value: string | number | Uint8Array, byteOffset?: number, encoding?: BufferEncoding): boolean;
export declare function bufferStripEndWithByBuffer(buf: Buffer, value: Uint8Array, byteOffset?: number, encoding?: BufferEncoding): Buffer;
export declare function bufferStripEndWith(buf: Buffer, value: string | number | Uint8Array, byteOffset?: number, encoding?: BufferEncoding): Buffer;
export declare function bufferIndexWithByBuffer(buf: Buffer, value: Uint8Array, byteOffset: number, encoding?: BufferEncoding): boolean;
export declare function bufferIndexWith(buf: Buffer, value: string | number | Uint8Array, byteOffset: number, encoding?: BufferEncoding): boolean;
export declare function splitBufferByBuffer(buffer: Buffer, value: Uint8Array): Buffer[];
