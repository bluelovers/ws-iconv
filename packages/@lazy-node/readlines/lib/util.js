"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitBufferByBuffer = exports.bufferIndexWith = exports.bufferIndexWithByBuffer = exports.bufferStripEndWith = exports.bufferStripEndWithByBuffer = exports.bufferEndWith = exports.bufferEndWithByBuffer = exports.handleOptionNewLineCharacter = void 0;
function handleOptionNewLineCharacter(newLineCharacter) {
    if (!newLineCharacter) {
        return [10 /* EnumNewLineCharacter.LF */];
    }
    else if (typeof newLineCharacter !== 'number') {
        let ls = [];
        for (let i = 0; i < newLineCharacter.length; i++) {
            let c = newLineCharacter[i];
            if (typeof c !== 'number') {
                ls.push(c.charCodeAt(0));
            }
            else {
                ls.push(c);
            }
        }
        return ls;
    }
    return [newLineCharacter];
}
exports.handleOptionNewLineCharacter = handleOptionNewLineCharacter;
function bufferEndWithByBuffer(buf, value, byteOffset, encoding) {
    let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
    return i !== -1 && (i + value.length) === buf.length;
}
exports.bufferEndWithByBuffer = bufferEndWithByBuffer;
function bufferEndWith(buf, value, byteOffset, encoding) {
    if (typeof value === 'number') {
        return buf[buf.length - 1] === value;
    }
    let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
    return i !== -1 && (i + value.length) === buf.length;
}
exports.bufferEndWith = bufferEndWith;
function bufferStripEndWithByBuffer(buf, value, byteOffset, encoding) {
    let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
    if (i !== -1 && (i + value.length) === buf.length) {
        return buf.slice(0, i);
    }
    return buf;
}
exports.bufferStripEndWithByBuffer = bufferStripEndWithByBuffer;
function bufferStripEndWith(buf, value, byteOffset, encoding) {
    if (typeof value === 'number') {
        if (buf[buf.length - 1] === value) {
            return buf.slice(0, buf.length - 1);
        }
    }
    else {
        let i = buf.lastIndexOf(value, byteOffset !== null && byteOffset !== void 0 ? byteOffset : (0 - value.length), encoding);
        if (i !== -1 && (i + value.length) === buf.length) {
            /*
            console.dir({
                i,
                vl: value.length,
                bl: buf.length,

                v: value.toString(),
                b: buf.toString(),
                b2: buf.slice(0, i).toString(),
            })
             */
            return buf.slice(0, i);
        }
    }
    return buf;
}
exports.bufferStripEndWith = bufferStripEndWith;
function bufferIndexWithByBuffer(buf, value, byteOffset, encoding) {
    let i = buf.indexOf(value, byteOffset, encoding);
    return i !== -1 && i === byteOffset;
}
exports.bufferIndexWithByBuffer = bufferIndexWithByBuffer;
function bufferIndexWith(buf, value, byteOffset, encoding) {
    if (typeof value === 'number') {
        return buf[byteOffset] === value;
    }
    let i = buf.indexOf(value, byteOffset, encoding);
    return i !== -1 && i === byteOffset;
}
exports.bufferIndexWith = bufferIndexWith;
function splitBufferByBuffer(buffer, value) {
    const lines = [];
    let bufferPosition = 0;
    let lastNewLineBufferPosition = 0;
    while (true) {
        let line;
        let bufferPositionValue = buffer[bufferPosition];
        if (bufferIndexWithByBuffer(buffer, value, bufferPosition++)) {
            line = buffer.slice(lastNewLineBufferPosition, bufferPosition += (value.length - 1));
            lines.push(line);
            lastNewLineBufferPosition = bufferPosition;
        }
        else if (bufferPositionValue === undefined) {
            break;
        }
    }
    let leftovers = buffer.slice(lastNewLineBufferPosition, bufferPosition);
    if (leftovers.length) {
        lines.push(leftovers);
    }
    return lines;
}
exports.splitBufferByBuffer = splitBufferByBuffer;
//# sourceMappingURL=util.js.map