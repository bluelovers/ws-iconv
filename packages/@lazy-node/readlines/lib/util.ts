import { INewLineCharacter, EnumNewLineCharacter } from './types';

export function handleOptionNewLineCharacter(newLineCharacter?: INewLineCharacter): number[]
{
	if (!newLineCharacter)
	{
		return [EnumNewLineCharacter.LF];
	}
	else if (typeof newLineCharacter !== 'number')
	{
		let ls = [] as number[]

		for (let i = 0; i < newLineCharacter.length; i++)
		{
			let c = newLineCharacter[i];

			if (typeof c !== 'number')
			{
				ls.push(c.charCodeAt(0))
			}
			else
			{
				ls.push(c)
			}
		}

		return ls
	}

	return [newLineCharacter]
}

export function bufferEndWithByBuffer(buf: Buffer,
	value: Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

	return i !== -1 && (i + value.length) === buf.length
}

export function bufferEndWith(buf: Buffer,
	value: string | number | Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	if (typeof value === 'number')
	{
		return buf[buf.length - 1] === value
	}

	let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

	return i !== -1 && (i + value.length) === buf.length
}

export function bufferStripEndWithByBuffer(buf: Buffer,
	value: Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

	if (i !== -1 && (i + value.length) === buf.length)
	{
		return buf.slice(0, i)
	}

	return buf
}

export function bufferStripEndWith(buf: Buffer,
	value: string | number | Uint8Array,
	byteOffset?: number,
	encoding?: BufferEncoding,
)
{
	if (typeof value === 'number')
	{
		if (buf[buf.length - 1] === value)
		{
			return buf.slice(0, buf.length - 1);
		}
	}
	else
	{
		let i = buf.lastIndexOf(value, byteOffset ?? (0 - value.length), encoding)

		if (i !== -1 && (i + value.length) === buf.length)
		{
			/*
			console.dir({
				i,
				vl: value.length,
				bl: buf.length,

				v: value.toString(),
				b: buf.toString(),
				b2: buf.slice(0, i).toString(),
			})
			 */

			return buf.slice(0, i)
		}
	}

	return buf
}

export function bufferIndexWithByBuffer(buf: Buffer,
	value: Uint8Array,
	byteOffset: number,
	encoding?: BufferEncoding,)
{
	let i = buf.indexOf(value, byteOffset, encoding)

	return i !== -1 && i === byteOffset
}

export function bufferIndexWith(buf: Buffer,
	value: string | number | Uint8Array,
	byteOffset: number,
	encoding?: BufferEncoding,)
{
	if (typeof value === 'number')
	{
		return buf[byteOffset] === value
	}

	let i = buf.indexOf(value, byteOffset, encoding)

	return i !== -1 && i === byteOffset
}

export function splitBufferByBuffer(buffer: Buffer, value: Uint8Array)
{
	const lines: Buffer[] = [];
	let bufferPosition = 0;

	let lastNewLineBufferPosition = 0;
	while (true)
	{
		let line: Buffer;
		let bufferPositionValue = buffer[bufferPosition];

		if (bufferIndexWithByBuffer(buffer, value, bufferPosition++))
		{
			line = buffer.slice(lastNewLineBufferPosition, bufferPosition += (value.length - 1));
			lines.push(line);
			lastNewLineBufferPosition = bufferPosition;
		}
		else if (bufferPositionValue === undefined)
		{
			break;
		}
	}

	let leftovers = buffer.slice(lastNewLineBufferPosition, bufferPosition);
	if (leftovers.length)
	{
		lines.push(leftovers);
	}

	return lines;
}
