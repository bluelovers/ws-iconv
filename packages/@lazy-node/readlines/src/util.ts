import { INewLineCharacter } from './types';
import { EnumNewLineCharacter } from './index';

export function handleOptionNewLineCharacter(newLineCharacter?: INewLineCharacter): number[]
{
	if (!newLineCharacter)
	{
		return [EnumNewLineCharacter.LF];
	}
	else if (typeof newLineCharacter !== 'number')
	{
		const ls = [] as number[];

		for (let i = 0; i < newLineCharacter.length; i++)
		{
			const c = newLineCharacter[i];

			if (typeof c !== 'number')
			{
				ls.push(c.charCodeAt(0))
			}
			else
			{
				ls.push(c)
			}
		}

		return ls
	}

	return [newLineCharacter]
}

