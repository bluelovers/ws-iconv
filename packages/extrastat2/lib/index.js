"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.statExtra = void 0;
const promises_1 = require("fs/promises");
const options_1 = require("./options");
const path_1 = require("path");
function statExtra(pathname, options) {
    const resolvedpath = (0, path_1.resolve)(pathname);
    const parsedPath = (0, path_1.parse)(resolvedpath);
    return (0, promises_1.stat)(resolvedpath)
        // @ts-ignore
        .then((statData) => {
        statData.parsed = parsedPath;
        return Promise
            .all((0, options_1.getOptionResolvers)(options).map(fn => fn(parsedPath, statData)))
            .then(() => {
            return statData;
        });
    });
}
exports.statExtra = statExtra;
exports.default = statExtra;
//# sourceMappingURL=index.js.map