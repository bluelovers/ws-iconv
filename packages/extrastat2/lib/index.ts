import { stat } from 'fs/promises';
import { getOptionResolvers} from './options';
import { resolve, parse } from 'path';
import { IStatsExtra, IOptions } from './types';

export function statExtra(pathname: string, options?: IOptions)
{
	const resolvedpath = resolve(pathname);
	const parsedPath = parse(resolvedpath)

	return stat(resolvedpath)
		// @ts-ignore
		.then((statData: IStatsExtra) => {

			statData.parsed = parsedPath;

			return Promise
				.all(getOptionResolvers(options).map(fn=> fn(parsedPath, statData)))
				.then(() => {
					return statData
				})
		})
	;
}

export default statExtra
