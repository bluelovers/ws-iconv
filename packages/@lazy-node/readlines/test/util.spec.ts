import { newContent } from './util';
import { splitBufferByBuffer, bufferStripEndWithByBuffer } from '../lib/util';

test(`splitBufferByBuffer`, () =>
{
	let buf = Buffer.from(newContent(`\n`))

	let c = Buffer.from(`\n`);

	let actual = splitBufferByBuffer(buf, c);
	let expected;

	let lines = actual.map(v => v.toString());

	expect(lines.join(``)).toStrictEqual(buf.toString());

	expect(lines).toMatchSnapshot();
	expect(lines.join(``)).toMatchSnapshot();

	expect(actual).toMatchSnapshot();

	expect(actual.map(v => bufferStripEndWithByBuffer(v, c))).toMatchSnapshot();
	expect(actual.map(v => bufferStripEndWithByBuffer(v, c)).join(`\n`)).toStrictEqual(buf.toString());

});

