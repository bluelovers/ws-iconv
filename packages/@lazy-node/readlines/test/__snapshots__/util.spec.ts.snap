// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitBufferByBuffer 1`] = `
Array [
  "0
",
  "1
",
  "2
",
  "3
",
  "4
",
  "5
",
  "6
",
  "7
",
  "8
",
  "9",
]
`;

exports[`splitBufferByBuffer 2`] = `
"0
1
2
3
4
5
6
7
8
9"
`;

exports[`splitBufferByBuffer 3`] = `
Array [
  Object {
    "data": Array [
      48,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      49,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      50,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      51,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      52,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      53,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      54,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      55,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      56,
      10,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      57,
    ],
    "type": "Buffer",
  },
]
`;

exports[`splitBufferByBuffer 4`] = `
Array [
  Object {
    "data": Array [
      48,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      49,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      50,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      51,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      52,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      53,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      54,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      55,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      56,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      57,
    ],
    "type": "Buffer",
  },
]
`;
